### PR TITLE
AP_NavEKF2: provide reasons for init failures

### PR DESCRIPTION
**Pull request by kd0aij**
replaces #12705
Instead of "AHRS unhealthy" , reports the reason for init failure when the core pointer is null.
Cases handled: 
1) insufficient memory
2) EK2_enable is false
3) no IMUs available
4) EK2_IMU_MASK is zero
5) core setup failed

### CODE REVIEW

File: libraries/AP_NavEKF2/AP_NavEKF2.h [hunk 1]
```diff
@@ -493,6 +493,28 @@ class NavEKF2 {
...
+        NO_ENABLE, 
+        NO_IMUS, 
+        NO_MASK, 
+        NO_ALLOC,
+        UNKNOWN,
+        NUM_INIT_FAILURES
+    };
+    // initialization failure reasons
+    const char* init_failure_reason[int(INIT_FAILURES::NUM_INIT_FAILURES)] {
+        "OK",
```
Comment (rmackay9) [inline libraries/AP_NavEKF2/AP_NavEKF2.h]: I think we may need to prepend "EKF2:" or something similar or the pre-arm failure will be, "PreArm: OK" or "PreArm:memory allocation failed".

Maybe the first option should be "Unknown"?

File: libraries/AP_NavEKF2/AP_NavEKF2.h [hunk 2]
```diff
@@ -493,6 +493,28 @@ class NavEKF2 {
     // time of last lane switch
     uint32_t lastLaneSwitch_ms;
 
+    enum class INIT_FAILURES {
```
Comment (rmackay9) [inline libraries/AP_NavEKF2/AP_NavEKF2.h]: It seems in the EKF the convention for enums is to use camel case.  So perhaps it should be "InitFailures"

**Comment by rmackay9** [comment]
I quickly tested this on a CubeOrange and got the nice "PreArm: EK2_IMU_MASK is zero" message after setting EK2_IMU_MASK = 0 which is more helpful than the "no EK2 cores" we see in master or "AHRS not healthy" that we see in Copter-4.0.0-rc1.

**Comment by kd0aij** [comment]
I like camelCase! But there seems to be no consistency between camelCase and camel_case in the EKF code, though the style guide seems to call for camel_case.
I also dislike ALL_CAPS for constants, but I don't think our style guide mentions this.

```
arm throttle
MANUAL> APM: PreArm: EKF2: EK2_IMU_MASK is zero
Got MAVLink msg: COMMAND_ACK {command : 400, result : 4}
Flight battery 100 percent
param set EK2_IMU_MASK 3
MANUAL> APM: EKF2 IMU0 initial yaw alignment complete
APM: EKF2 IMU1 initial yaw alignment complete
APM: EKF2 IMU0 tilt alignment complete
APM: EKF2 IMU1 tilt alignment complete
APM: EKF2 IMU0 origin set
APM: EKF2 IMU1 origin set
param set EK2_ENABLE 0
MANUAL> reboot
...
Init ArduPlane V4.1.0dev (14cf898c)

Free RAM: 131072
load_all took 11661us
0 0 0 AP_Logger_File: buffer size=51200
APM: Ground start
APM: Skipping INS calibration
APM: Calibrating barometer
INITIALISING> Mode INITIALISING
APM: Barometer 1 calibration complete
APM: Airspeed calibration started
APM: Ground start complete
MANUAL> APM: Throttle failsafe off
...
Mode MANUAL
APM: Airspeed 1 calibrated
arm throttle
MANUAL> APM: PreArm: EKF2: EK2_enable is false
```

**Comment by rmackay9** [comment]
ok, re my camel case comment I was just referring to the name of the enum.. i would have left the other names as they were.  Anyway.. let's see if @tridge or @priseborough has time to review this.  If not I might merge it in any case.

**Comment by kd0aij** [comment]
@rmackay9 Thanks for reviewing. I'm afraid these additions would have to be duplicated for EKF3, the way the code is structured. Would be nice if this could move up a level (somewhere) instead.

**Comment by rmackay9** [comment]
@kd0aij, yes, they'll need to be duplicated for EKF3 but that's often the case so I wouldn't worry.

**Comment by kd0aij** [comment]
Report failures in core setup and finish initialization in the case that not all succeed.

**Comment by kd0aij** [comment]
add another message for not all cores are allocated and fail prearm
and consolidate the 2 insufficient memory errors

File: libraries/AP_NavEKF2/AP_NavEKF2.h (hunk 1 above)
Comment (kd0aij) [reply libraries/AP_NavEKF2/AP_NavEKF2.h]: done

**Comment by tridge** [comment]
pending CI

**Comment by rmackay9** [comment]
@kd0aij, one small issue I've found with this change is that it shows the last init failure instead of the first.  This means that if it runs out of memory it displays, "EKF2: EK2_enable is false" instead of "EKF2: insufficient memory available".  I think the reason is that InitialiseFilter is run more than once.  Perhaps we should only set the initFailure to a new value if it's InitFailures::UNKNOWN

**Comment by kd0aij** [comment]
That's my fault for not testing all scenarios :(
Perhaps it would be best not to force the enable to zero though?
Then the correct error should persist, unless more memory magically appeared.

**Comment by rmackay9** [comment]
As discussed on mumble, I totally agree that setting the enable to zero is not good.  I really think we should avoid this pattern... we don't do this kind of thing too often but I think it's nearly always a bad thing to do.

**Comment by kd0aij** [comment]
fixed in https://github.com/ArduPilot/ardupilot/pull/12799

**Pull request merged**